include/supercluster.hpp

```diff
@@ -286,41 +288,63 @@ class Supercluster {
 
                 p.visited = true;
 
-                auto num_points = p.num_points;
-                point<double> weight = p.pos * double(num_points);
-
-                auto clusterProperties = p.properties ? *p.properties : property_map{};
-                std::uint32_t id = static_cast<std::uint32_t>((i << 5) + (zoom + 1));
-
-                // find all nearby points
+                const auto num_points_origin = p.num_points;
+                auto num_points = num_points_origin;
+                auto cluster_size = previous.clusters.size();
+                // count the number of points in a potential cluster
                 previous.tree.within(p.pos.x, p.pos.y, r, [&](const auto &neighbor_id) {
-                    assert(neighbor_id < previous.clusters.size());
-                    auto &b = previous.clusters[neighbor_id];
-
+                    assert(neighbor_id < cluster_size);
+                    const auto &b = previous.clusters[neighbor_id];
                     // filter out neighbors that are already processed
-                    if (b.visited) {
-                        return;
-                    }
-
-                    b.visited = true;
-                    b.parent_id = id;
-
-                    // accumulate coordinates for calculating weighted center
-                    weight += b.pos * double(b.num_points);
-                    num_points += b.num_points;
-
-                    if (options_.reduce && b.properties) {
-                        // apply reduce function to update clusterProperites
-                        options_.reduce(clusterProperties, *b.properties);
+                    if (!b.visited) {
+                        num_points += b.num_points;
                     }
                 });
 
-                if (num_points == 1) {
-                    clusters.emplace_back(weight, 1, p.id, clusterProperties);
-                } else {
+                auto clusterProperties = p.properties ? *p.properties : property_map{};
```

Comment:
`clusterProperties` only mutated if `num_points >= options_.minPoints` and if we have reduce function. Could we share `p.properties` between clusters if no mutation is required?